openml/datasets/dataset.py

```diff
@@ -158,41 +159,106 @@ def __init__(self, name, description, format=None,
+                                    "there and can be read.", self.data_file)
+                    raise e
+
+                ARFF_DTYPES_TO_PD_DTYPE = {
+                    'INTEGER': 'integer',
+                    'REAL': 'floating',
+                    'NUMERIC': 'floating',
+                    'STRING': 'string'
+                }
+                attribute_dtype = {}
+                attribute_names = []
+                categories_names = {}
+                categorical = []
+                for name, type_ in data['attributes']:
+                    # if the feature is nominal and the a sparse matrix is
+                    # requested, the categories need to be numeric
+                    if (isinstance(type_, list) and
+                            self.format.lower() == 'sparse_arff'):
+                        try:
+                            np.array(type_, dtype=np.float32)
+                        except ValueError:
+                            raise ValueError(
+                                "Categorical data needs to be numeric when "
+                                "using sparse ARFF."
+                            )
+                    # string can only be supported with pandas DataFrame
+                    elif (type_ == 'STRING' and
+                          self.format.lower() == 'sparse_arff'):
+                        raise ValueError(
+                            "Dataset containing strings is not supported "
+                            "with sparse ARFF."
+                        )
+
+                    # infer the dtype from the ARFF header
+                    if isinstance(type_, list):
+                        categorical.append(True)
+                        categories_names[name] = type_
+                        if set(['True', 'False']) == set(type_):
+                            attribute_dtype[name] = 'boolean'
+                        else:
+                            attribute_dtype[name] = 'categorical'
                     else:
-                        raise Exception()
+                        categorical.append(False)
+                        attribute_dtype[name] = ARFF_DTYPES_TO_PD_DTYPE[type_]
+                    attribute_names.append(name)
+
+                if self.format.lower() == 'sparse_arff':
+                    X = data['data']
+                    X_shape = (max(X[1]) + 1, max(X[2]) + 1)
+                    X = scipy.sparse.coo_matrix(
+                        (X[0], (X[1], X[2])), shape=X_shape, dtype=np.float32)
+                    X = X.tocsr()
+
+                elif self.format.lower() == 'arff':
+                    X = pd.DataFrame(data['data'], columns=attribute_names)
+                    X_null_count = X.isnull().sum()
+
+                    for column_name in X.columns:
+                        if attribute_dtype[column_name] == 'boolean':
+                            X[column_name] = self._create_column_bool(
```

Comment:
I'm afraid that this could become extremely costly for large datasets with a large amount of columns. Do you think it would be feasible (in terms of memory) to store all columns in a list first and then concatenate them at the end of the function?

Reply:
I would rely on @jorisvandenbossche for the internal of pandas. But I don't think that it will be more costly. I am unsure that it will be more costly.

Reply:
I don't know about the internals of pandas, but I made the experience that changing the columns inplace can be quite expensive:
```
import pandas as pd
import numpy as np
array = np.random.random(size=(1000, 10000))
df = pd.DataFrame(array)

def cast_array(df):
    df_copy = df.copy()
    for name in df_copy.columns:
        df_copy[name] = df_copy[name].astype(float)
    return df_copy

def cast_array_2(df):
    df_copy = df.copy()
    columns = []
    for name in df_copy.columns:
        columns.append(df_copy[name].astype(float))
    return pd.concat(columns, axis=1)

df_copy_1 = cast_array(df)
df_copy_2 = cast_array_2(df)
print(np.all(df_copy_1 == df_copy_2))
%timeit cast_array(df)
%timeit cast_array_2(df)
```
results in
```
## -- End pasted text --
True
1.6 s ± 75.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.06 s ± 75.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
The second one is actually what `df.astype(..)` is doing under the hood if you provide a dict of {col: new_dtype} (this is indeed more efficient than actually astyping single columns on the full dataframe one by one (the first function)).

Reply:
OK. Make sense.